File: backend/rust/src/domain/value/iban.rs
// IBAN value object — validation lives here; an `IBAN` is always well-formed.

use core::fmt;
use core::str::FromStr;
use rand::Rng;
use crate::domain::errors::DomainError;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct IBAN(String);

impl IBAN {

    pub fn generate() -> Result<Self, DomainError> {
        let mut rng = rand::thread_rng();

        // 1. Build Romanian BBAN: bank + branch + 12 random digits
        let country   = "RO";
        let bank_code = "GNTL";
        let branch    = "0000";

        let account: String = (0..12)
            .map(|_| rng.gen_range(b'0'..=b'9') as char)
            .collect();

        let bban = format!("{}{}{}", bank_code, branch, account);

        // 2. Compute ISO 13616 check digits: 98 - mod97(BBAN + country + "00")
        let raw = format!("{}{}00", bban, country);
        let check_digits = 98 - Self::mod97_str(&raw);

        // 3. Assemble final 24-char RO IBAN
        let iban_str = format!("{}{:02}{}", country, check_digits, bban);

        Ok(IBAN(iban_str))
    }

    fn mod97_str(s: &str) -> u64 {
        // Expand letters to digits (A=10..Z=35), then mod 97 one digit at a time
        s.chars()
            .flat_map(|c| {
                if c.is_ascii_alphabetic() {
                    let n = c.to_ascii_uppercase() as u32 - b'A' as u32 + 10;
                    n.to_string().chars().collect::<Vec<_>>()
                } else {
                    vec![c]
                }
            })
            .fold(0u64, |acc, c| {
                (acc * 10 + c.to_digit(10).unwrap() as u64) % 97
            })
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn into_string(self) -> String {
        self.0
    }
}

impl fmt::Display for IBAN {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl FromStr for IBAN {
    type Err = DomainError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // 1. Normalize: trim, strip internal whitespace, uppercase
        let raw = s.trim();
        if raw.is_empty() {
            return Err(DomainError::validation("IBAN must not be empty"));
        }

        let iban: String = raw.chars().filter(|c| !c.is_whitespace()).collect::<String>().to_ascii_uppercase();

        // 2. Enforce fixed RO IBAN shape
        if iban.len() != 24 { return Err(DomainError::validation("Invalid IBAN length")); }
        if !iban.starts_with("RO") { return Err(DomainError::validation("Invalid country code")); }

        Ok(IBAN(iban))
    }
}

impl TryFrom<String> for IBAN {
    type Error = DomainError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        IBAN::from_str(&value)
    }
}

impl From<IBAN> for String {
    fn from(value: IBAN) -> Self {
        value.0
    }
}

File: backend/c/domain/value/iban.c
#include "../include/iban.h"

#include <ctype.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

const char *iban_as_cstr(const IBAN *iban) {
    return iban ? iban->value : "";
}

bool iban_try_create(const char *raw, IBAN *out, DomainError *err) {
    if (!raw || !out) {
        if (err) *err = domain_error_validation("IBAN must not be null");
        return false;
    }

    // Allow one extra char so over-length input is detected (not silently truncated).
    char buffer[IBAN_MAX_LEN + 2];
    size_t j = 0;
    for (size_t i = 0; raw[i] != '\0' && j <= IBAN_MAX_LEN; ++i) {
        char c = raw[i];
        if (!isspace((unsigned char)c)) {
            buffer[j++] = (char)toupper((unsigned char)c);
        }
    }
    buffer[j] = '\0';

    if (j == 0) {
        if (err) *err = domain_error_validation("IBAN must not be empty");
        return false;
    }
    if (j != 24) {
        if (err) *err = domain_error_validation("Invalid IBAN length");
        return false;
    }
    if (buffer[0] != 'R' || buffer[1] != 'O') {
        if (err) *err = domain_error_validation("Invalid country code");
        return false;
    }

    strcpy(out->value, buffer);
    if (err) *err = domain_error_ok();
    return true;
}

// IBAN check digits: treat BBAN+country as a big integer mod 97 (ISO 13616).
// Letters expand to two digits (A→10 … Z→35); we fold digit-by-digit to stay in uint64.
static uint64_t mod97_str(const char *s) {
    uint64_t acc = 0;
    for (; *s; ++s) {
        char c = *s;
        if (c >= 'A' && c <= 'Z') {
            // Letters expand to two digits (A→10 … Z→35).
            int n = (c - 'A') + 10;
            acc = (acc * 10 + (n / 10)) % 97;
            acc = (acc * 10 + (n % 10)) % 97;
        } else {
            acc = (acc * 10 + (c - '0')) % 97;
        }
    }
    return acc;
}

bool iban_generate(IBAN *out, DomainError *err) {
    if (!out) {
        if (err) *err = domain_error_validation("IBAN output must not be null");
        return false;
    }

    static int seeded = 0;
    if (!seeded) { srand((unsigned)time(NULL)); seeded = 1; }

    const char *country   = "RO";
    const char *bank_code = "GNTL";
    const char *branch    = "0000";

    char account[13];
    for (int i = 0; i < 12; ++i)
        account[i] = (char)('0' + rand() % 10);
    account[12] = '\0';

    char bban[21];
    snprintf(bban, sizeof bban, "%s%s%s", bank_code, branch, account);

    char raw[32];
    // Rearrange to BBAN+country+"00" so mod97 yields the check digits.
    snprintf(raw, sizeof raw, "%s%s00", bban, country);

    int check_digits = (int)(98 - mod97_str(raw));

    char iban_str[IBAN_MAX_LEN + 1];
    snprintf(iban_str, sizeof iban_str, "%s%02d%s", country, check_digits, bban);

    strcpy(out->value, iban_str);
    if (err) *err = domain_error_ok();
    return true;
}
